shared/src/ledger/protocol/transactions/votes.rs

```diff
@@ -121,3 +121,104 @@ where
     storage.write(&keys.voting_power(), &tally.voting_power.try_to_vec()?)?;
     Ok(())
 }
+
+/// Deterministically constructs a [`Votes`] map from a set of validator
+/// addresses and the block heights they signed something at. We arbitrarily
+/// take the earliest block height for each validator address encountered.
+pub fn dedupe(signers: BTreeSet<(Address, BlockHeight)>) -> Votes {
+    signers.into_iter().rev().collect()
```

Comment:
I had some involved chained iterators here then realized we can just use the natural `signers.into_iter().rev().collect()`.

As I understand it, it shouldn't make a difference whether we use the earliest or latest vote by a validator for an event within a digest (especially since we don't guarantee that we will take the globally earliest/latest vote in any case) - the important thing is that we only use exactly one vote per validator per event all in all. I am guessing validators should prefer their earlier votes to be used, since in an ideal world (i.e. ABCI++) their first vote for an event would always be recorded and they wouldn't have to submit multiple votes at all.